File: YeetMacro2/Services/Scripts/KonosubaFD/watchAdStamina.cs
using Microsoft.Extensions.Logging;
namespace YeetMacro2.Services.Scripts.KonosubaFD;
public partial class KonosubaFDScripts
{
    public string watchAdStamina()
    {
        // patterns["stamina"]["adNotification"]
        //var done = false;
        //var loopPatterns = new PatternNode[] { patterns["titles"]["home"], patterns["stamina"]["prompt"]["recoverStamina"] };
        //while (macroService.IsRunning && !done)
        //{
        //    var result = macroService.PollPattern(loopPatterns);
        //    switch (result.Path)
        //    {
        //        case "titles.home":
        //            logger.LogInformation("watchAdStamina: stamina ad");
        //            macroService.ClickPattern(patterns["stamina"]["add"]);
        //            break;
        //        case "stamina.prompt.recoverStamina":
        //            logger.LogInformation("watchAdStamina: check for stamina adNotification");
        //            var staminaAdNotificationResult = macroService.FindPattern(patterns["stamina"]["adNotification"]);
        //            logger.LogInformation("staminaAdNotificationResult.IsSuccess: " + staminaAdNotificationResult.IsSuccess);
        //            if (!staminaAdNotificationResult.IsSuccess)
        //            {
        //                logger.LogInformation("watchAdStamina: stamina ad notification not detected");
        //                var watchAdResult = macroService.PollPattern(patterns["stamina"]["watchAd"], new PollPatternFindOptions() { DoClick = true, IntervalDelayMs: 1000, PredicatePattern = new PatternNode[] { patterns["prompt"]["dailyRewardLimit"], patterns["stamina"]["adNotification"] } });
        //            if (watchAdResult.PredicatePath == "prompt.dailyRewardLimit")
        //            {
        //                logger.LogInformation("watchAdStamina: daily reward limit");
        //                macroService.PollPattern(patterns["ad"]["prompt"]["ok"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["stamina"]["prompt"]["recoverStamina"] });
        //                done = true;
        //                break;
        //            }
        //            break;
        //    }

        //    logger.LogInformation("watchAdStamina: watching ad");
        //    logger.LogInformation("watchAdStamina: poll stamina.adNotification");
        //    macroService.PollPattern(patterns["stamina"]["adNotification"], new PollPatternFindOptions() { DoClick = true, ClickPattern = patterns["ad"]["prompt"]["ok"], PredicatePattern = patterns["ad"]["done"] });
        //    Sleep(1_000);
        //    logger.LogInformation("watchAdStamina: poll ad.done");
        //    macroService.PollPattern(patterns["ad"]["done"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["ad"]["prompt"]["ok"] });
        //    Sleep(1_000);
        //    logger.LogInformation("watchAdStamina: poll ad.prompt.ok 2");
        //    macroService.PollPattern(patterns["ad"]["prompt"]["ok"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["stamina"]["add"] });

        //    done = true;
        //    break;
        //}

        //Sleep(1_000);
        //}
        //logger.LogInformation("Done...");
        return String.Empty;
    }
}

File: YeetMacro2/Resources/Raw/MacroSets/konosubaFD/scripts/_watchAdStamina.js
// @isFavorite
// @position=-100
const loopPatterns = [patterns.titles.home, patterns.stamina.prompt.recoverStamina];
const daily = dailyManager.GetCurrentDaily();
const adObjectScale = settings.watchAdStamina.adObjectScale.Value;
const adExitPattern = adObjectScale === 1 ? patterns.ad.exit : macroService.ClonePattern(patterns.ad.exit, { Scale: adObjectScale });
const adExitInstallPattern = adObjectScale === 1 ? patterns.ad.exitInstall : macroService.ClonePattern(patterns.ad.exitInstall, { Scale: adObjectScale });
const adCancelPattern = adObjectScale === 1 ? patterns.ad.cancel : macroService.ClonePattern(patterns.ad.cancel, { Scale: adObjectScale });
const soundLeftPattern = adObjectScale === 1 ? patterns.ad.soundLeft : macroService.ClonePattern(patterns.ad.soundLeft, { Scale: adObjectScale });
const userClickPattern = macroService.ClonePattern(settings.watchAdStamina.userClickPattern.Value, {
	Path: 'settings.watchAdStamina.userClickPattern'
});

if (daily.watchAdStamina.done.IsChecked) {
	return "Script already completed. Uncheck done to override daily flag.";
}

while (macroService.IsRunning) {
	const result = macroService.PollPattern(loopPatterns);
	switch (result.Path) {
		case 'titles.home':
			logger.info('watchAdStamina: stamina ad');
			macroService.ClickPattern(patterns.stamina.add);
			break;
		case 'stamina.prompt.recoverStamina':
			logger.info('watchAdStamina: check if disabled');
			const watchResult = macroService.PollPattern(patterns.ad.stamina.watch, { TimeoutMs: 3_000 });
			if (!watchResult.IsSuccess) {
				return;
			}

			logger.info('watchAdStamina: watching ad');
			macroService.PollPattern(patterns.ad.stamina.watch, { DoClick: true, PredicatePattern: patterns.ad.prompt.ok });
			macroService.PollPattern(patterns.ad.prompt.ok, { DoClick: true, InversePredicatePattern: patterns.ad.prompt.ok });
			macroService.PollPattern(patterns.ad.prompt.ok, {
				ClickPattern: [adExitInstallPattern, adExitPattern, adCancelPattern, soundLeftPattern, userClickPattern]
			});
			macroService.PollPattern(patterns.ad.prompt.ok, { DoClick: true, PredicatePattern: [patterns.titles.home, patterns.stamina.prompt.recoverStamina] });
			if (macroService.IsRunning) {
				daily.watchAdStamina.count.Count++;
				if (daily.watchAdStamina.count.Count >= 3) {
					daily.watchAdStamina.done.IsChecked = true;
					return;
				}
			}
			break;
	}

	sleep(1_000);
}